### PR TITLE
sql: Add system table join_tokens, and create_join_tokens() builtin function

### PR DESCRIPTION
**Pull request by itsbilal**
This change adds a new system table, `join_tokens`, for the
exclusive use of storing join tokens. This is necessary as
we need guaranteed at-most-once semantics with these, which
transactions give us pretty easily. A related migration is also added
to create said table

This change also adds a new builtin function, `crdb_internal.create_join_token()`
that creates and persists a join token in that table.

Currently, there's no mechanism to remove expired join tokens.

See RFC #51991. Part of #60632.

Release note (general change): Add `crdb_internal.create_join_token()`
sql builtin function to create join tokens for use when joining
new nodes to a secure cluster. This functionality is hidden behind
a feature flag.

### CODE REVIEW

**Comment by itsbilal** [comment]
There are many SQL bits here so it'd rather have this get some more SQL eyes. That said, the first and third commits might have an impact on some of your work @aaron-crl , as it moves around some `JoinToken` stuff.

**Review by knz — COMMENTED** [review]
1st commit LGTM

2nd commit not quite so. I think this work does not justify a fully fledged new SQL statement. (Moreover, CREATE is someone earmarked for DDL / SQL metaschema changes.)  I would recommend a SQL built-in function instead. We already have `crdb_internal.create_tenant` and I believe we can follow this model here.

3rd commit LGTM

4th commit code OK, but when you connect the create function to the cert manager, don't do this in the `sql/sem/builtins` package. Instead add a function in the sql package and make it an interface. Again, you can follow the example of `create_tenant`.

5th commit LGTM

> Reviewed 3 of 3 files at r1, 11 of 11 files at r2, 4 of 4 files at r3, 9 of 9 files at r4, 5 of 5 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62053#-:-MWdDbk76slQhrWiWa65:bhksjv2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aaron-crl and @dt)

**Comment by itsbilal** [comment]
TFTR! I've moved over to a sql builtin, and updated the PR accordingly. Also folded the other PR into this one.

**Review by knz — APPROVED** [review]
:lgtm: but I think you need to revisit some tests that assert table IDs.

> Reviewed 26 of 26 files at r6, 4 of 4 files at r7, 4 of 4 files at r8, 13 of 13 files at r9, 5 of 5 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62053#-:-MWxvBNlB2yms-MNPeLe:b-5236rn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aaron-crl, @dt, and @itsbilal)

___
*[pkg/sql/join_token_test.go, line 87 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62053#-MWxuQLV5LwfIzMaiqb6:-MWxuQLWCOvTZNOLJtPI:b-h15v7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/23156ae4f93cbddf887f1029258821dccac7422e/pkg/sql/join_token_test.go#L87)):*
> ```Go
> 	defer s.Stopper().Stop(context.Background())
> 
> 	rows, err := sqldb.Query("CREATE JOINTOKEN;")
> ```

2nd commit: CREATE JOINTOKEN doesn't seem relevant to the PR any more. Maybe you can copy the query from the 4th commit into it.

**Review by itsbilal — COMMENTED** [review]
TFTR! Updated the failing tests, and fixed lint issues too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62053#-:-MWybGdU8thAIrjNQAEB:bkswhyh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aaron-crl, @dt, and @knz)

___
*[pkg/sql/join_token_test.go, line 87 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/62053#-MWxuQLV5LwfIzMaiqb6:-MWybLiABymt3UeOlga1:bqe02c9) ([raw file](https://github.com/cockroachdb/cockroach/blob/23156ae4f93cbddf887f1029258821dccac7422e/pkg/sql/join_token_test.go#L87)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

2nd commit: CREATE JOINTOKEN doesn't seem relevant to the PR any more. Maybe you can copy the query from the 4th commit into it.
</blockquote></details>

Done. Moved this entire file to the 4th commit to reduce confusion, instead of having a half-working test.

**Comment by itsbilal** [comment]
bors r=knz

File: pkg/sql/catalog/systemschema/system.go [hunk 1]
```diff
@@ -1756,6 +1764,49 @@ var (
 		FormatVersion:  descpb.InterleavedFormatVersion,
 		NextMutationID: 1,
 	})
+
+	// MigrationsTable is the descriptor for the migrations table. It stores facts
```
Comment (irfansharif) [inline pkg/sql/catalog/systemschema/system.go:1768]: PS I think this comment could do with fixing, it's referring to the table before.

**Pull request merged**